File: src/test/java/io/tripled/social/client/presentation/ReadEvalPrintLoopTest.java
package io.tripled.social.client.presentation;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.junit.jupiter.api.Assertions.assertThrows;

class ReadEvalPrintLoopTest {

  private ReadEvalPrintLoop repl;
  private TestInput input;
  private TestOutput output;

  @BeforeEach
  void setUpReplWithEchoCommand() {
    input = new TestInput();
    output = new TestOutput();
    repl = new ReadEvalPrintLoop(input, output, null, null, null, null);
  }

  @Test
  void throwsExceptionWithNullInput() {
    assertThrows(NullPointerException.class, () -> {
      new ReadEvalPrintLoop(null, output, null, null, null, null);
    });
  }

  @Test
  void throwsExceptionWithNullOutput() {
    assertThrows(NullPointerException.class, () -> {
      new ReadEvalPrintLoop(new TestInput(), null, null, null, null, null);
    });
  }

  @Test
  void quits() {
    input.addInput("\\q");

    repl.run();

    assertThat(repl.isStopped(), is(true));
    output.assertIsEmpty();
  }

}

File: src/test/kotlin/io/tripled/social/client/presentation/ReadEvalPrintLoopTest.kt
package io.tripled.social.client.presentation

import io.tripled.social.client.application.SpiedFollowUserUseCase
import io.tripled.social.client.application.SpiedPostMessageUseCase
import io.tripled.social.client.application.SpiedReadMessagesUseCase
import io.tripled.social.client.application.SpiedReadWallUseCase
import io.tripled.social.client.presentation.controller.FollowController
import io.tripled.social.client.presentation.controller.PostController
import io.tripled.social.client.presentation.controller.ReadController
import io.tripled.social.client.presentation.controller.WallController
import org.hamcrest.MatcherAssert
import org.hamcrest.Matchers
import org.junit.jupiter.api.Test

internal class ReadEvalPrintLoopTest {
    private val input: TestInput = TestInput()
    private val output: TestOutput = TestOutput()
    private val repl: ReadEvalPrintLoop = ReadEvalPrintLoop(
        input, output,
        PostController(SpiedPostMessageUseCase()),
        ReadController(SpiedReadMessagesUseCase()),
        WallController(SpiedReadWallUseCase()),
        FollowController(SpiedFollowUserUseCase())
    )

    @Test
    fun quits() {
        input.addInput("\\q")

        repl.run()

        MatcherAssert.assertThat<Boolean?>(repl.isStopped, Matchers.`is`<Boolean?>(true))
        output.assertIsEmpty()
    }
}
